programs/host/frontend.c

```diff
@@ -0,0 +1,943 @@
+        }
+        break;
+
+    case MBEDTLS_SERIALIZE_FUNCTION_FOPEN:
+        {
+            char * path = NULL;
+            char * mode = NULL;
+            int32_t file_id;
+
+            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+            mode = item_buffer( inputs[0] );
+            path = item_buffer( inputs[1] );
+            DBG( "open file [%s] mode [%s]", path, mode);
+            file_id = alloc_file_context();
+            DBG( "allocated file id [%d]", file_id);
+            if ( file_id != -1 )
+            {
+                FILE * file = fopen( path, mode );
+                if ( file != NULL )
+                {
+                    files[file_id - 1].file = file;
+                    ALLOC_OUTPUT( 0, sizeof (int32_t) );
+                    set_item_int32( outputs[0], file_id );
+                    ret = 0;
+                }
+                else
+                {
+                    DBG( "fopen: error = %s", strerror( errno ) );
+                    free_file_context( file_id );
+                }
+            }
+        }
+        break;
+    case MBEDTLS_SERIALIZE_FUNCTION_FREAD:
+        {
+            int32_t file_id, size;
+            FILE * file = NULL;
+
+            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+            size = item_int32( inputs[0] );
+            file_id = item_int32( inputs[1] );
+            file = (FILE *)get_file_from_id( file_id );
+            if ( file )
+            {
+                ALLOC_OUTPUT( 0, size );
+                ret = fread( item_buffer( outputs[0] ), 1, size, file );
+                if ( ret >= 0 )
+                {
+                    outputs[0]->size = ret;
+                    ret = 0;
+                }
+            }
+        }
+        break;
+    case MBEDTLS_SERIALIZE_FUNCTION_FGETS:
+        {
+            int32_t file_id, size;
+            FILE * file = NULL;
+
+            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+            size = item_int32( inputs[0] );
+            file_id = item_int32( inputs[1] );
+            file = (FILE *)get_file_from_id( file_id );
+            if ( file )
+            {
+                char * s = NULL;
+                ALLOC_OUTPUT( 0, size );
+                s = fgets( item_buffer( outputs[0] ), size, file );
+                if ( s != NULL )
+                {
+                    outputs[0]->size = strlen( s ) + 1;
+                    ret = 0;
+                }
+            }
+        }
+        break;
+    case MBEDTLS_SERIALIZE_FUNCTION_FWRITE:
+        {
+            int32_t file_id;
+            FILE * file = NULL;
+
+            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+            file_id = item_int32( inputs[1] );
+            file = (FILE *)get_file_from_id( file_id );
+            if ( file )
+            {
+                ret = fwrite( item_buffer( inputs[0] ), 1, inputs[0]->size, file );
+                if ( ret >= 0 )
+                {
+                    ALLOC_OUTPUT( 0, sizeof( int32_t ) );
```

Comment:
Call `ALLOC_OUTPUT` before doing any I/O. The way the code is, if the allocation fails, the write has been done but the application thinks it failed.

Same thing in other functions.